### PR TITLE
[DI] Documenting Abstract Bundle and Extension

### LINKED ISSUE #16669: [HttpKernel] Simplifying Bundle/Extension config definition
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#43701
| PR author(s) | @yceruto
| Merged in    | 6.1

### LINKED ISSUE #16696: [Notifier] [OvhCloud] Add `no_stop_clause` to DSN
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#45951
| PR author(s) | @alamirault
| Merged in    | 6.1

### PR DESCRIPTION
**Pull request by yceruto**
Fixes #16669
PR https://github.com/symfony/symfony/pull/43701

ping @Nyholm @wouterj as you were interested in this feature/doc. Any suggestion to improve the wording is more than welcome!

---

I also updated other places to be aligned with the best practices doc.
I didn't remove the current convention to create bundles/extensions/configurations because they are still valid until 6.4 becomes the last LTS available. However, it's ok for apps using AbstractExtension.

### CODE REVIEW

File: bundles.rst
Comment (yceruto) [inline bundles.rst:51]: Leftover of the previous bundle-based app approach

File: bundles.rst
Comment (yceruto) [inline bundles.rst]: I would like to change the default convention to not be forced to adjust the bundle path. The recommended bundle directory structure is already supported since Symfony 4.4, so we could take advantage of this new class to change the default path without BC breaks.

File: bundles.rst
Comment (yceruto) [inline bundles.rst:115]: maybe it's more evident if we show the composer.json and autoload(-dev) mapping on this page.

File: bundles.rst
Comment (yceruto) [inline bundles.rst]: maybe we should change this path to `src/` only and explain what it contains, not only controllers

File: bundles.rst
Comment (yceruto) [inline bundles.rst:92]: I propose to remove this entry from this very first moment, as the current text says, it's not necessary. After, in another doc we're mentioning this dir for extension convention

File: bundles.rst
Comment (yceruto) [inline bundles.rst:101]: Updating according to best practices for bundle directory structure

**Comment by yceruto** [comment]
About CI jobs, Is there any way to know which code block is failing? and how these failures can be fixed?:
```
bundles/prepend_extension.rst ✘
  174: Please reorder the use statements alphabetically
   ->  and defining the :method:`Symfony\\Component\\HttpKernel\\Bundle\\AbstractBundle::prependExtension` method::

bundles/extension.rst ✘
  160: Please reorder the use statements alphabetically
   ->  and defining the :method:`Symfony\\Component\\HttpKernel\\Bundle\\AbstractBundle::loadExtension` method::

Warning: ] Found "2" invalid files!                                             

Error: Please reorder the use statements alphabetically
Error: Please reorder the use statements alphabetically
```

File: bundles/configuration.rst
Comment (HeahDude) [inline bundles/configuration.rst]: When I read this I understand we're talking about the "old" fashion way to do it, whereas the following block is showing the new way. I would replace this `where` by `now`, or reword this sentence altogether.

File: bundles/configuration.rst
Comment (HeahDude) [inline bundles/configuration.rst]:
```suggestion
    The "configure()" method is called only at compile time.
```
?

File: bundles/extension.rst
Comment (HeahDude) [inline bundles/extension.rst]: This comment is about the feature itself (maybe too late now), but how getting the builder here can be useful? What can be done with it that cannot be done with the configurator?

File: bundles/extension.rst
Comment (HeahDude) [inline bundles/extension.rst]: Did we somehow forget to mention https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/HttpKernel/DependencyInjection/ConfigurableExtension.php#L41?

File: bundles/extension.rst
Comment (HeahDude) [inline bundles/extension.rst]: `compile time` as well?

Comment (HeahDude) [inline bundles/extension.rst]:
```suggestion
    The "loadExtension()", as the "load()" method, are called only at compile time.
```

File: configuration/micro_kernel_trait.rst
Comment (HeahDude) [inline configuration/micro_kernel_trait.rst]: Changes in this file do not seem related, 🤔

File: configuration/micro_kernel_trait.rst
Comment (HeahDude) [inline configuration/micro_kernel_trait.rst:226]: Missing note about adding this abstract class in 6.1

File: configuration/micro_kernel_trait.rst
Comment (HeahDude) [inline configuration/micro_kernel_trait.rst]:
```suggestion
                $container->set('foo_service', new \stdClass());
```

File: configuration/micro_kernel_trait.rst
Comment (HeahDude) [inline configuration/micro_kernel_trait.rst]: Missing `'dev' === $this->environment` check

**Review by HeahDude — COMMENTED** [review]
Thanks for this feature and its docs :)

File: configuration/micro_kernel_trait.rst
Comment (HeahDude) [reply configuration/micro_kernel_trait.rst:311]: https://github.com/symfony/symfony-docs/issues/15223#issuecomment-1126023101 😞

File: bundles/extension.rst
Comment (yceruto) [reply bundles/extension.rst]: The `ContainerBuilder` contains still useful methods that are not included in the `ContainerConfigurator`, for example, `removeDefinition()`, `hasDefinition()`, `registerForAutoconfiguration()`, `prependExtensionConfig()`, `getExtensionConfig()`, etc. (the last two useful in the new `prependExtension()`) 

In this method, we used to import the service definitions from an external file, and then we can remove some if desired using the builder instance. That's why we kept both.

File: bundles/extension.rst
Comment (yceruto) [reply bundles/extension.rst]: The `ConfigurableExtension` class is not the one used in this feature, in fact, this new feature replaces that class, and I guess we can deprecate it in future versions. https://github.com/symfony/symfony/pull/43701#discussion_r825310122

File: configuration/micro_kernel_trait.rst
Comment (yceruto) [reply configuration/micro_kernel_trait.rst]: You're right, it's not related and I'm going to revert it to not add noise.

File: configuration/micro_kernel_trait.rst
Comment (yceruto) [reply configuration/micro_kernel_trait.rst]: Reverted according to the previous comment.

File: bundles/extension.rst
Comment (HeahDude) [reply bundles/extension.rst]: I see, thank you for your explanation :).

File: bundles.rst
Comment (yceruto) [reply bundles.rst:115]: Done.

File: bundles.rst
Comment (yceruto) [reply bundles.rst]: Done.

File: bundles.rst
Comment (apfelbox) [reply bundles.rst]: Yes please! 🙌 
Was always a pet peeve

Comment (yceruto) [reply bundles.rst]: Done in https://github.com/symfony/symfony/pull/46385 (waiting for review/merge)

Comment (yceruto) [reply bundles.rst]: Note/Code removed. Now when your bundle extends from `AbstractBundle` Symfony assumes you're using the modern directory structure.

File: bundles.rst
Comment (HeahDude) [inline bundles.rst]: I would move that "if" in a warning/info note below this one, wdyt?

File: bundles.rst
Comment (HeahDude) [inline bundles.rst:77]: To be backported in 4.4?

Comment (yceruto) [reply bundles.rst:77]: Yup, the whole bundle doc should be updated since 4.4.

Comment (yceruto) [reply bundles.rst:77]: I will create a separate PR for that.

Comment (wouterj) [reply bundles.rst:77]: I don't think backporting is needed. We often don't do big rewrites on old versions. We simply don't have much resources on documentation, so getting current in good shape is hard enough already.
Besides that, nobody should start writing a new bundle in their Symfony 4.4 application and the current 4.4 doc already served 4.4 users for 2.5 years, we'll manage that the final 6 months as well :)

File: configuration/micro_kernel_trait.rst
Comment (wouterj) [inline configuration/micro_kernel_trait.rst:215]: To be honest, I'm not really sure if this is a good place to document `AbstractExtension`. What is its use-case when using MicroKernelTrait?

If we want to keep it, I would suggest moving the changes in this document to a new section below this one. Currently, I'm afraid the doc can make people think that they must use the app extension (or `build()` method) when using the micro kernel.

File: configuration/micro_kernel_trait.rst
Comment (wouterj) [inline configuration/micro_kernel_trait.rst]: Please revert this change. We should make the change on 5.4 and we decided to always show both annotation and attribute for now (using a `.. configuration-block::`).

**Review by wouterj — APPROVED** [review]
Hi @yceruto! Thanks again for all these amazing docs.

I've pushed a commit to your branch doing a bit more radical rewrite. Please have a look at the changes. This PR is ready to merge if you approve them imho :)
As I said on Slack, we might want to rewrite all bundle docs in the future, to swap the order (first `AbstractBundle`, then Extension+Configuration). Yet, the current changes are perfect for the 6.1 release already I think.

File: bundles/configuration.rst
Comment (yceruto) [inline bundles/configuration.rst]:
```suggestion
    configuration definition from one or more files::
```

File: bundles/extension.rst
Comment (yceruto) [inline bundles/extension.rst]: Do you pretend to upgrade the directory structure? In that case, the sentences above should be updated too.

File: configuration/micro_kernel_trait.rst
Comment (yceruto) [reply configuration/micro_kernel_trait.rst:215]: > To be honest, I'm not really sure if this is a good place to document AbstractExtension. What is its use-case when using MicroKernelTrait?

Apps on bundle-less approach still need to define custom configuration, and it's only possible through a new extension, even if the kernel class can implement the ExtensionInterface + Configuration it's better now by creating a new extension class and extending from AbstractExtension.

> If we want to keep it, I would suggest moving the changes in this document to a new section below this one. Currently, I'm afraid the doc can make people think that they must use the app extension (or build() method) when using the micro kernel.

I think you're right, it could be better in a new section with the proper explanation.

File: bundles/extension.rst
Comment (wouterj) [reply bundles/extension.rst]: Oh, you're right. Yes, let's use the new bundle structure everywhere in the 6.1 docs :)

File: configuration/micro_kernel_trait.rst
Comment (yceruto) [reply configuration/micro_kernel_trait.rst]: Reverted.

File: bundles/extension.rst
Comment (yceruto) [reply bundles/extension.rst]: Updated, also the path here is missing one `/../`as the extension class in this case is placed in `src/DependencyInjection/`

**Comment by yceruto** [comment]
@wouterj thanks a lot for your improvements, it looks better now 🙏

**Comment by javiereguiluz** [comment]
This was finally merged 🎉 

Yonel, thanks again for such a nice contribution! Thanks to reviewers too!

**Pull request merged**
